thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -1246,10 +1246,18 @@ namespace diskann {
                                indices.data(), distances.data(), beam_width,
                                false, stats, bitset_view);
       for (_u32 i = 0; i < l_search; i++) {
-        if (distances[i] > (float) range || indices[i] == -1) {
-          res_count = i;
-          break;
-        } else if (i == l_search - 1)
+        if (metric== diskann::Metric::INNER_PRODUCT) {
+          if (distances[i] < (float) range || indices[i] == -1) {
+            res_count = i;
+            break;
+          }
+        } else {
+          if (distances[i] > (float) range || indices[i] == -1) {
+            res_count = i;
+            break;
+          }
+        }
+        if (i == l_search - 1)
```

Comment:
Two problems:
1. this embedded `if` logic's readability is too low.
2. we need to check `indces[i] == -1` first. Since `distances[i] == -1` doesn't means `indices[i] != -1`.

Suggest to make it 
```
      if (indices[i] == -1) {
        res_count = i;
        break;
      }
      bool out_of_range = metric == diskann::Metric::INNER_PRODUCT
                              ? distances[i] < (float) range
                              : distances[i] > (float) range;
      if (out_of_range) {
        res_count = i;
        break;
      }
      if (i == l_search - 1) {
        res_count = l_search;
      }
```